pyomo/contrib/matching/interface.py

```diff
@@ -0,0 +1,79 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.collections import ComponentMap
+from pyomo.contrib.matching.maximum_matching import maximum_matching
+from pyomo.contrib.matching.block_triangularize import block_triangularize
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+
+class IncidenceGraphInterface(object):
+    """
+    The purpose of this class is to allow the user to easily
+    analyze graphs of variables and contraints in a Pyomo
+    model without constructing multiple PyomoNLPs.
+    """
+
+    def __init__(self, model):
+        self.nlp = PyomoNLP(model)
+
```

Comment:
I think I would like to see another method here to reduce the amount of duplicated code:

```python
def get_incidence_matrix(self, obj, variables, constraints):
    if isinstance(model, PyomoNLP):
        return self.get_pyomonlp_incidence_matrix(obj, variables, constraints)
    else:
        return self.get_pyomo_incidence_matrix(obj, variables, constraints)

def get_pyomonlp_incidence_matrix(self, nlp, variables, constraints):
    if variables is None:
        variables = nlp.get_pyomo_variables()
    if constraints is None:
        constraints = nlp.get_pyomo_constraints()

    for comp in variables+constraints:
        if comp.is_indexed():
            raise ValueError(
                        "Variables and constraints must be unindexed "
                        "ComponentData objects. Got %s, which is indexed."
                        % comp.name
                        )
    matrix = nlp.extract_submatrix_jacobian(variables, constraints)
    return variables, constraints, matrix

def get_pyomo_incidence_matrix(self, model, variables, constraints):
    _check = []
    if variables is None:
        variables = list(model.component_data_objects(Var))
    else:
        _check.extend(variables)
    if constraints is None:
        constraints = list(model.component_data_objects(Constraint, active=True))
    else:
        _check.extend(constraints)

    for comp in _check:
        if comp.is_indexed():
            raise ValueError(
                        "Variables and constraints must be unindexed "
                        "ComponentData objects. Got %s, which is indexed."
                        % comp.name
                        )
    _varID = ComponentMap((v,i) for i,v in enumerate(variables))
    _rows = []
    _cols = []
    for i, c in enumerate(constraints):
        _cols.extend(_varID[v] for v in identify_variables(c.body) if v in _varID)
        _rows.extend([i]*(len(_cols)-len(_rows)))
    matrix = scipy.sparse.coo_matrix( ([1]*len(_rows), (_rows, _cols)), 
            shape=(len(constraints), len(variables)) )
    return variables, constraints, matrix

def maximum_matching(self, variables=None, constraints=None):
    var, con, matrix = self.get_incidence_matrix(self.nlp, variables, constraints)
    matching = maximum_matching(matrix.tocoo())
    # Matching maps row (constraint) indices to column (variable) indices
    return ComponentMap((con[i], varj]) for i, j in matching.items())

def block_triangularize(self, variables=None, constraints=None):
    var, con, matrix = self.get_incidence_matrix(self.nlp, variables, constraints)
    ....
```


```

Reply:
See comment below. I added a method to get rid of the repeated code, but I believe "structural incidence matrix" code should probably be added as a function independent of this class. I can do that as part of this PR.